File: ios/loveaceios/ViewModel/TeacherEvaluationViewModel.swift
import Foundation

@MainActor @Observable
final class TeacherEvaluationViewModel {
    var isLoading = false
    var isEvaluating = false
    var courses: [TeacherEvaluationCourse] = []
    var tasks: [TeacherEvaluationTask] = []
    var logs: [String] = []
    var error: String?
    var isEvaluationClosed = false
    var lastResult: TeacherEvaluationBatchResult?
    var evaluationStrategy: EvaluationStrategy = .smart

    private var service: TeacherEvaluationService?
    private var launcherTask: Task<Void, Never>?
    private var activeTasks: [Task<TeacherEvaluationResult, Never>] = []
    private let launchIntervalSeconds = 6
    private let countdownSeconds = 140

    var pendingCourses: [TeacherEvaluationCourse] { courses.filter { !$0.isEvaluated } }
    var evaluatedCourses: [TeacherEvaluationCourse] { courses.filter(\.isEvaluated) }
    var totalCount: Int { courses.count }
    var pendingCount: Int { pendingCourses.count }
    var evaluatedCount: Int { evaluatedCourses.count }
    var completedTaskCount: Int { tasks.filter(\.isFinished).count }
    var successfulTaskCount: Int { tasks.filter(\.isSuccess).count }
    var progress: Double { tasks.isEmpty ? 0 : Double(completedTaskCount) / Double(tasks.count) }

    func initialize(service: TeacherEvaluationService) {
        self.service = service
    }

    func loadCourses() {
        guard let service else { return }
        Task {
            isLoading = true
            error = nil
            isEvaluationClosed = false
            let result = await service.fetchCourses()
            if result.success, let data = result.data {
                courses = data
            } else if Self.isClosedMessage(result.error) {
                courses = []
                tasks = []
                isEvaluationClosed = true
            } else {
                error = result.error ?? "加载课程失败"
            }
            isLoading = false
        }
    }

    func startConcurrentEvaluation() {
        guard !isEvaluating else { return }
        guard service != nil else {
            error = "评教服务未初始化"
            return
        }
        guard !isEvaluationClosed else {
            error = "评价暂未开启"
            return
        }
        let pending = pendingCourses
        guard !pending.isEmpty else {
            error = "没有待评课程"
            return
        }

        isEvaluating = true
        error = nil
        lastResult = nil
        logs.removeAll()
        tasks = pending.enumerated().map { index, course in
            TeacherEvaluationTask(taskId: index + 1, course: course)
        }
        addLog("开始并发批量评教，共 \(pending.count) 门课程")

        launcherTask = Task { [weak self] in
            guard let self else { return }
            await self.runConcurrentEvaluation()
        }
    }

    func cancelEvaluation() {
        guard isEvaluating else { return }
        launcherTask?.cancel()
        activeTasks.forEach { $0.cancel() }
        markUnfinishedTasksAsCancelled()
        addLog("已请求中断评教任务")
        isEvaluating = false
    }

    func clearResult() {
        lastResult = nil
    }

    private func runConcurrentEvaluation() async {
        guard !tasks.isEmpty else {
            isEvaluating = false
            return
        }

        let startTime = Date()
        var runners: [Task<TeacherEvaluationResult, Never>] = []

        for task in tasks {
            if Task.isCancelled { break }
            if task.taskId > 1 {
                addLog("等待 \(launchIntervalSeconds) 秒后启动下一个任务（\(task.taskId)/\(tasks.count)）")
                do {
                    try await Task.sleep(for: .seconds(launchIntervalSeconds))
                } catch {
                    break
                }
            }

            addLog("启动任务 \(task.taskId)：\(task.course.displayName)")
            let runner = Task { [weak self] in
                guard let self else {
                    return TeacherEvaluationResult(course: task.course, success: false, errorMessage: "任务已释放")
                }
                return await self.executeTask(taskId: task.taskId)
            }
            activeTasks.append(runner)
            runners.append(runner)
        }

        if Task.isCancelled {
            activeTasks.forEach { $0.cancel() }
            markUnfinishedTasksAsCancelled()
            isEvaluating = false
            activeTasks.removeAll()
            launcherTask = nil
            return
        }

        addLog("所有任务已启动，等待完成...")
        var results: [TeacherEvaluationResult] = []
        for runner in runners {
            results.append(await runner.value)
        }

        let success = results.filter(\.success).count
        let failed = results.count - success
        lastResult = TeacherEvaluationBatchResult(
            total: results.count,
            success: success,
            failed: failed,
            results: results,
            duration: Date().timeIntervalSince(startTime)
        )

        if failed > 0 {
            addLog("并发评教完成：成功 \(success)/\(results.count)，失败 \(failed)")
        } else {
            addLog("并发评教完成：全部 \(results.count) 门课程成功")
        }

        await refreshCoursesAfterEvaluation()
        isEvaluating = false
        activeTasks.removeAll()
        launcherTask = nil
    }

    private func executeTask(taskId: Int) async -> TeacherEvaluationResult {
        guard let service, let task = tasks.first(where: { $0.taskId == taskId }) else {
            return TeacherEvaluationResult(course: TeacherEvaluationCourse(
                legacyId: "", name: "未知课程", teacher: "", evaluatedPeople: "", evaluatedPeopleNumber: "",
                coureSequenceNumber: "", evaluationContentNumber: "", questionnaireCode: "", questionnaireName: "", isEvaluated: false
            ), success: false, errorMessage: "任务不存在")
        }

        let course = task.course
        updateTask(taskId, status: .preparing, message: "访问评价页面", startTime: Date())
        addLog("任务 \(taskId) [\(course.displayName)]：访问评价页面")

        if Task.isCancelled { return failTask(taskId, course: course, message: "已取消") }

        let prepared = await service.prepareEvaluation(course: course, totalCourses: tasks.count, strategy: evaluationStrategy)
        guard prepared.success, let formData = prepared.data else {
            return failTask(taskId, course: course, message: prepared.error ?? "无法访问评价页面")
        }

        updateTask(taskId, status: .preparing, message: "解析问卷")
        addLog("任务 \(taskId) [\(course.displayName)]：解析问卷")
        updateTask(taskId, status: .preparing, message: "生成答案")
        addLog("任务 \(taskId) [\(course.displayName)]：生成答案")

        updateTask(taskId, status: .countdown, message: "等待提交", countdownRemaining: countdownSeconds, countdownTotal: countdownSeconds)
        addLog("任务 \(taskId) [\(course.displayName)]：开始等待 \(countdownSeconds) 秒")
        for remaining in stride(from: countdownSeconds, through: 1, by: -1) {
            if Task.isCancelled { return failTask(taskId, course: course, message: "已取消") }
            updateTask(taskId, countdownRemaining: remaining, countdownTotal: countdownSeconds)
            do {
                try await Task.sleep(for: .seconds(1))
            } catch {
                return failTask(taskId, course: course, message: "已取消")
            }
        }

        updateTask(taskId, status: .submitting, message: "提交评价", countdownRemaining: 0, countdownTotal: countdownSeconds)
        addLog("任务 \(taskId) [\(course.displayName)]：提交评价")
        let submit = await service.submitEvaluation(formData: formData)
        guard submit.success else {
            return failTask(taskId, course: course, message: submit.error ?? "提交失败")
        }

        updateTask(taskId, status: .verifying, message: "验证结果")
        addLog("任务 \(taskId) [\(course.displayName)]：验证结果")
        let verify = await service.fetchCourses()
        guard verify.success, let updatedCourses = verify.data else {
            return failTask(taskId, course: course, message: verify.error ?? "无法验证结果")
        }
        let updatedCourse = updatedCourses.first { course.matches($0) }
        guard updatedCourse?.isEvaluated == true else {
            return failTask(taskId, course: course, message: "评教未生效，服务器未确认")
        }

        updateTask(taskId, status: .completed, message: "完成", endTime: Date())
        addLog("任务 \(taskId) [\(course.displayName)]：评教完成")
        return TeacherEvaluationResult(course: course, success: true)
    }

    private func refreshCoursesAfterEvaluation() async {
        guard let service else { return }
        addLog("刷新课程列表...")
        let result = await service.fetchCourses()
        if result.success, let data = result.data {
            courses = data
            addLog("课程列表已更新")
        } else if let error = result.error {
            addLog("课程列表刷新失败：\(error)")
        }
    }

    private func failTask(_ taskId: Int, course: TeacherEvaluationCourse, message: String) -> TeacherEvaluationResult {
        updateTask(taskId, status: .failed, message: "失败", errorMessage: message, endTime: Date())
        addLog("任务 \(taskId) [\(course.displayName)]：\(message)")
        return TeacherEvaluationResult(course: course, success: false, errorMessage: message)
    }

    private func updateTask(
        _ taskId: Int,
        status: TeacherEvaluationTaskStatus? = nil,
        message: String? = nil,
        errorMessage: String? = nil,
        countdownRemaining: Int? = nil,
        countdownTotal: Int? = nil,
        startTime: Date? = nil,
        endTime: Date? = nil
    ) {
        guard let index = tasks.firstIndex(where: { $0.taskId == taskId }) else { return }
        if let status { tasks[index].status = status }
        if let message { tasks[index].statusMessage = message }
        if let errorMessage { tasks[index].errorMessage = errorMessage }
        if let countdownRemaining { tasks[index].countdownRemaining = countdownRemaining }
        if let countdownTotal { tasks[index].countdownTotal = countdownTotal }
        if let startTime { tasks[index].startTime = startTime }
        if let endTime { tasks[index].endTime = endTime }
    }

    private func markUnfinishedTasksAsCancelled() {
        for index in tasks.indices where !tasks[index].isFinished {
            tasks[index].status = .failed
            tasks[index].statusMessage = "已取消"
            tasks[index].errorMessage = "用户已中断任务"
            tasks[index].endTime = Date()
        }
    }

    private func addLog(_ message: String) {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss"
        logs.append("[\(formatter.string(from: Date()))] \(message)")
        if logs.count > 120 { logs.removeFirst(logs.count - 120) }
    }

    private static func isClosedMessage(_ message: String?) -> Bool {
        guard let message else { return false }
        return message.contains("评价暂未开启") || message.contains("评估开关已关闭")
    }
}

File: android/app/src/main/java/tech/loveace/appv3/ui/viewmodel/TeacherEvaluationViewModel.kt
package tech.loveace.appv3.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import tech.loveace.appv3.data.model.TeacherEvaluationCourse
import tech.loveace.appv3.data.service.TeacherEvaluationService

enum class TeacherEvaluationTaskStatus { Pending, Preparing, Waiting, Submitting, Verifying, Success, Failed, Cancelled }

data class TeacherEvaluationTaskState(
    val course: TeacherEvaluationCourse,
    val status: TeacherEvaluationTaskStatus = TeacherEvaluationTaskStatus.Pending,
    val message: String = "等待开始",
    val countdownSeconds: Int = 0,
)

data class TeacherEvaluationUiState(
    val isLoading: Boolean = false,
    val isRunning: Boolean = false,
    val isClosed: Boolean = false,
    val closedMessage: String = "",
    val indexToken: String = "",
    val courses: List<TeacherEvaluationCourse> = emptyList(),
    val tasks: List<TeacherEvaluationTaskState> = emptyList(),
    val logs: List<String> = emptyList(),
    val error: String? = null,
    val evaluationStrategy: EvaluationStrategy = EvaluationStrategy.Smart,
) {
    val pendingCourses: List<TeacherEvaluationCourse>
        get() = courses.filter { !it.isEvaluated }

    val evaluatedCount: Int
        get() = courses.count { it.isEvaluated }
}

class TeacherEvaluationViewModel : ViewModel() {
    private var service: TeacherEvaluationService? = null
    private var schedulerJob: Job? = null

    private val _uiState = MutableStateFlow(TeacherEvaluationUiState())
    val uiState: StateFlow<TeacherEvaluationUiState> = _uiState.asStateFlow()

    fun init(service: TeacherEvaluationService) {
        this.service = service
    }

    fun load() {
        val svc = service ?: return
        stop(resetRunning = false)
        viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isLoading = true,
                    isRunning = false,
                    isClosed = false,
                    closedMessage = "",
                    error = null,
                    tasks = emptyList(),
                )
            }
            val result = svc.loadCourses()
            val data = result.data
            if (!result.success || data == null) {
                val error = result.error ?: "获取评教课程失败"
                _uiState.update { it.copy(isLoading = false, error = error) }
                addLog(error)
                return@launch
            }
            if (data.isClosed) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        isClosed = true,
                        closedMessage = data.closedMessage.ifBlank { "评价暂未开启" },
                        indexToken = data.tokenValue,
                        courses = emptyList(),
                        tasks = emptyList(),
                    )
                }
                addLog("评价暂未开启")
                return@launch
            }
            _uiState.update {
                it.copy(
                    isLoading = false,
                    indexToken = data.tokenValue,
                    courses = data.courses,
                    tasks = emptyList(),
                    error = null,
                )
            }
            addLog("已加载 ${data.courses.size} 门课程，待评 ${data.courses.count { !it.isEvaluated }} 门")
        }
    }

    fun startBatch() {
        val svc = service ?: return
        val state = _uiState.value
        val pending = state.pendingCourses
        if (state.isRunning || pending.isEmpty() || state.indexToken.isBlank()) return

        stop(resetRunning = false)
        _uiState.update {
            it.copy(
                isRunning = true,
                tasks = pending.map { course -> TeacherEvaluationTaskState(course = course) },
                error = null,
            )
        }
        addLog("批量评教已启动，请保持 App 前台；已提交评价无法撤回")

        schedulerJob = viewModelScope.launch {
            val jobs = mutableListOf<Job>()
            try {
                pending.forEachIndexed { index, course ->
                    if (index > 0) delay(START_INTERVAL_MS)
                    jobs += launch { runTask(svc, course, pending.size, state.indexToken) }
                }
                jobs.joinAll()
                refreshAfterBatch(svc)
            } catch (_: CancellationException) {
                jobs.forEach { it.cancel() }
            } finally {
                _uiState.update { it.copy(isRunning = false) }
            }
        }
    }

    fun stop(resetRunning: Boolean = true) {
        schedulerJob?.cancel()
        schedulerJob = null
        if (resetRunning) {
            _uiState.update { state ->
                state.copy(
                    isRunning = false,
                    tasks = state.tasks.map { task ->
                        if (task.status.isActive()) {
                            task.copy(status = TeacherEvaluationTaskStatus.Cancelled, message = "已停止", countdownSeconds = 0)
                        } else task
                    },
                )
            }
            addLog("批量评教已停止；已提交的评价不会撤回")
        }
    }

    fun setStrategy(strategy: EvaluationStrategy) {
        if (_uiState.value.isRunning) return
        _uiState.update { it.copy(evaluationStrategy = strategy) }
    }

    private suspend fun runTask(
        service: TeacherEvaluationService,
        course: TeacherEvaluationCourse,
        pendingCount: Int,
        indexToken: String,
    ) {
        try {
            updateTask(course, TeacherEvaluationTaskStatus.Preparing, "正在访问评价页并生成表单")
            val prepare = service.prepareEvaluation(course, pendingCount, indexToken, state.evaluationStrategy)
            val prepared = prepare.data
            if (!prepare.success || prepared == null) {
                failTask(course, prepare.error ?: "准备评价表单失败")
                return
            }

            for (second in WAIT_BEFORE_SUBMIT_SECONDS downTo 1) {
                updateTask(course, TeacherEvaluationTaskStatus.Waiting, "等待提交", second)
                delay(1000)
            }

            updateTask(course, TeacherEvaluationTaskStatus.Submitting, "正在提交评价")
            val submit = service.submitEvaluation(prepared)
            val submitResult = submit.data
            if (!submit.success || submitResult?.success != true) {
                failTask(course, submitResult?.message ?: submit.error ?: "提交评价失败")
                return
            }

            updateTask(course, TeacherEvaluationTaskStatus.Verifying, "正在刷新课程列表验证")
            val verified = service.verifyCourseEvaluated(course)
            if (verified.success && verified.data == true) {
                updateTask(course, TeacherEvaluationTaskStatus.Success, "提交成功，服务器已确认")
                addLog("${course.displayName()} 提交成功")
            } else {
                failTask(course, verified.error ?: "评教未生效，服务器未确认")
            }
        } catch (_: CancellationException) {
            updateTask(course, TeacherEvaluationTaskStatus.Cancelled, "已取消", 0)
        }
    }

    private suspend fun refreshAfterBatch(service: TeacherEvaluationService) {
        val result = service.loadCourses()
        val data = result.data
        if (result.success && data != null && !data.isClosed) {
            _uiState.update { it.copy(courses = data.courses, indexToken = data.tokenValue) }
            addLog("批量任务结束，已刷新课程列表")
        } else if (result.error != null) {
            addLog("批量任务结束后刷新失败：${result.error}")
        }
    }

    private fun failTask(course: TeacherEvaluationCourse, message: String) {
        updateTask(course, TeacherEvaluationTaskStatus.Failed, message, 0)
        addLog("${course.displayName()} 失败：$message")
    }

    private fun updateTask(
        course: TeacherEvaluationCourse,
        status: TeacherEvaluationTaskStatus,
        message: String,
        countdown: Int = 0,
    ) {
        _uiState.update { state ->
            state.copy(
                tasks = state.tasks.map { task ->
                    if (task.course.matches(course)) {
                        task.copy(status = status, message = message, countdownSeconds = countdown)
                    } else task
                }
            )
        }
    }

    private fun addLog(message: String) {
        _uiState.update { it.copy(logs = (it.logs + message).takeLast(80)) }
    }

    override fun onCleared() {
        stop(resetRunning = false)
        super.onCleared()
    }

    private fun TeacherEvaluationTaskStatus.isActive(): Boolean = when (this) {
        TeacherEvaluationTaskStatus.Pending,
        TeacherEvaluationTaskStatus.Preparing,
        TeacherEvaluationTaskStatus.Waiting,
        TeacherEvaluationTaskStatus.Submitting,
        TeacherEvaluationTaskStatus.Verifying -> true
        TeacherEvaluationTaskStatus.Success,
        TeacherEvaluationTaskStatus.Failed,
        TeacherEvaluationTaskStatus.Cancelled -> false
    }

    private fun TeacherEvaluationCourse.displayName(): String = listOf(name, teacher)
        .filter { it.isNotBlank() }
        .joinToString(" / ")
        .ifBlank { displayId }

    companion object {
        private const val START_INTERVAL_MS = 6_000L
        private const val WAIT_BEFORE_SUBMIT_SECONDS = 140
    }
}
